### PR TITLE
MM-55476: Optimize PostStore.Get

### PR DESCRIPTION
**Pull request by agnivade**
While loading a thread, we were unnecessarily re-counting the replies
for all posts in a thread even if they would be the same number.

While this is needed in other queries where the posts can be from different
threads or they can be random ids, but to load a single post thread, there
is no need to recompute it again and again.

Therefore, we use a CTE to precompute the replycount and then just plug in
the value in the subsequent query. This gives an improvement in the query
plan as well:

OLD:
```
explain (analyze, buffers) SELECT p.id, p.rootid, p.createat, (SELECT count(*) FROM Posts WHERE Posts.RootId = (CASE WHEN p.RootId = '' THEN p.Id ELSE p.RootId END) AND Posts.DeleteAt = 0) as ReplyC
ount FROM Posts p WHERE (p.Id = 'h3cer597jb8abbcbitpghpomua' OR p.RootId = 'h3cer597jb8abbcbitpghpomua') AND p.DeleteAt = 0;
                                                                            QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Bitmap Heap Scan on posts p  (cost=45.39..1042149.61 rows=2748 width=49) (actual time=3.156..7906.215 rows=5353 loops=1)
   Recheck Cond: (((id)::text = 'h3cer597jb8abbcbitpghpomua'::text) OR (((rootid)::text = 'h3cer597jb8abbcbitpghpomua'::text) AND (deleteat = 0)))
   Filter: (deleteat = 0)
   Heap Blocks: exact=5308
   Buffers: shared hit=610244
   ->  BitmapOr  (cost=45.39..45.39 rows=2748 width=0) (actual time=0.918..0.920 rows=0 loops=1)
         Buffers: shared hit=47
         ->  Bitmap Index Scan on posts_pkey  (cost=0.00..1.68 rows=1 width=0) (actual time=0.028..0.028 rows=1 loops=1)
               Index Cond: ((id)::text = 'h3cer597jb8abbcbitpghpomua'::text)
               Buffers: shared hit=4
         ->  Bitmap Index Scan on idx_posts_root_id_delete_at  (cost=0.00..42.34 rows=2747 width=0) (actual time=0.889..0.890 rows=5352 loops=1)
               Index Cond: (((rootid)::text = 'h3cer597jb8abbcbitpghpomua'::text) AND (deleteat = 0))
               Buffers: shared hit=43
   SubPlan 1
     ->  Aggregate  (cost=378.10..378.11 rows=1 width=8) (actual time=1.474..1.474 rows=1 loops=5353)
           Buffers: shared hit=604889
           ->  Index Only Scan using idx_posts_root_id_delete_at on posts  (cost=0.57..343.85 rows=13699 width=0) (actual time=0.016..1.039 rows=5352 loops=5353)
                 Index Cond: ((rootid = (CASE WHEN ((p.rootid)::text = ''::text) THEN p.id ELSE p.rootid END)::text) AND (deleteat = 0))
                 Heap Fetches: 0
                 Buffers: shared hit=604889
 Planning Time: 0.194 ms
 Execution Time: 7906.846 ms
```

NEW:
```
explain analyze with replycount as (select count(*) as num from posts where rootid='h3cer597jb8abbcbitpghpomua' and deleteat=0)
select id, rootid, createat, replycount.num from posts, replycount where id='h3cer597jb8abbcbitpghpomua' or rootid='h3cer597jb8abbcbitpghpomua' and deleteat=0;
-------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Nested Loop  (cost=123.16..3215.48 rows=2748 width=49) (actual time=2.960..9.775 rows=5353 loops=1)
   ->  Aggregate  (cost=77.78..77.79 rows=1 width=8) (actual time=1.455..1.456 rows=1 loops=1)
         ->  Index Only Scan using idx_posts_root_id_delete_at on posts posts_1  (cost=0.57..70.91 rows=2747 width=0) (actual time=0.056..1.145 rows=5352 loops=1)
               Index Cond: ((rootid = 'h3cer597jb8abbcbitpghpomua'::text) AND (deleteat = 0))
               Heap Fetches: 0
   ->  Bitmap Heap Scan on posts  (cost=45.39..3110.20 rows=2748 width=41) (actual time=1.501..7.747 rows=5353 loops=1)
         Recheck Cond: (((id)::text = 'h3cer597jb8abbcbitpghpomua'::text) OR (((rootid)::text = 'h3cer597jb8abbcbitpghpomua'::text) AND (deleteat = 0)))
         Heap Blocks: exact=5308
         ->  BitmapOr  (cost=45.39..45.39 rows=2748 width=0) (actual time=0.797..0.798 rows=0 loops=1)
               ->  Bitmap Index Scan on posts_pkey  (cost=0.00..1.68 rows=1 width=0) (actual time=0.014..0.014 rows=1 loops=1)
                     Index Cond: ((id)::text = 'h3cer597jb8abbcbitpghpomua'::text)
               ->  Bitmap Index Scan on idx_posts_root_id_delete_at  (cost=0.00..42.34 rows=2747 width=0) (actual time=0.782..0.782 rows=5352 loops=1)
                     Index Cond: (((rootid)::text = 'h3cer597jb8abbcbitpghpomua'::text) AND (deleteat = 0))
 Planning Time: 0.220 ms
 Execution Time: 10.052 ms
(15 rows)
```

Observe the `loops=5353` in the first query, and `loops=1` in the next.

Fixes https://github.com/mattermost/mattermost/issues/25414
https://mattermost.atlassian.net/browse/MM-55476

```release-note
Optimize createPost performance
```

### CODE REVIEW

**Comment by agnivade** [comment]
This is one of those changes which are very specific and does not come up prominently on the load tests. I think mainly because we have low distribution of very long threads. Anyways, ran the tests and here are the relevant metrics. There seems to be an improvement if you squint, but it's close to calling it noise.

p99-store-times
p50-poststore get-times
p99-poststore get-times

File: server/channels/store/sqlstore/post_store.go [hunk 1]
```diff
@@ -716,14 +715,36 @@ func (s *SqlPostStore) Get(ctx context.Context, id string, opts model.GetPostsOp
...
+				Select("p.*, replycount.num as ReplyCount").
+				PrefixExpr(s.getQueryBuilder().
+					Select().
+					Prefix("WITH replycount as (").
+					Columns("count(*) as num").
+					From("posts").
+					Where(sq.And{
+						sq.Eq{"RootId": rootId},
+						sq.Eq{"DeleteAt": 0},
+					}).Suffix(")"),
```
Comment (agarciamontoro) [inline server/channels/store/sqlstore/post_store.go:739]: Super nit: I'd love to use only one style out of the two we see here:
```go
query.
    Where(sq.Eq{"x", x}).
    Where(sq.Eq{"y", y})
```
And

```go
query.Where(sq.And{
    sq.Eq{"x", x},
    sq.Eq{"y", y},
})
```

File: server/channels/store/sqlstore/post_store.go [hunk 2]
```diff
@@ -816,7 +837,7 @@ func (s *SqlPostStore) GetSingle(id string, inclDeleted bool) (*model.Post, erro
 		Where(sq.Eq{"p.Id": id})
 
 	replyCountSubQuery := s.getQueryBuilder().
-		Select("COUNT(Posts.Id)").
+		Select("COUNT(*)").
```
Comment (agarciamontoro) [inline server/channels/store/sqlstore/post_store.go:840]: A casual "let's optimize this now that we're around" :joy:

**Review by agarciamontoro — APPROVED** [review]
Amazing catch! A couple of questions for my own understanding, both about the other cases where you say that the subquery in the select part is indeed needed:
1. Which are those cases?
2. Would they benefit from a simple self join? Haven't tested that, just thinking out loud.

**Comment by agnivade** [comment]
1. Basically for most of the other methods, you just work with a list of ids. Those ids can be root ids or child ids. If we had stricter API assumptions around that, then we could have optimized. But since we don't know if they are root posts or replies, we have to compute that for every post id.
2. It won't have much benefit. It's not that you are repeating the ids. You'll have to calculate atleast once for each id anyways.

**Review by streamer45 — APPROVED** [review]
Good stuff :muscle: I guess it's true that the best performance optimization is removing work.

File: server/channels/store/sqlstore/post_store.go [hunk 3]
```diff
@@ -716,14 +715,36 @@ func (s *SqlPostStore) Get(ctx context.Context, id string, opts model.GetPostsOp
...
+					sq.Eq{"p.Id": rootId},
+					sq.Eq{"p.RootId": rootId},
+				}).
+				Where(sq.Eq{"p.DeleteAt": 0})
+		} else {
+			query = s.getQueryBuilder().
+				Select("p.*, replycount.num as ReplyCount").
+				PrefixExpr(s.getQueryBuilder().
+					Select().
+					Prefix("WITH replycount as (").
```
[the remark this answers is not in the document]
Comment (agnivade) [reply server/channels/store/sqlstore/post_store.go:733]: Haha yeah this is a weird one indeed.

**Comment by agnivade** [comment]
Hey @DHaussermann - would you mind giving this a smoke test with a focus around posts and threads? Thanks.

**Comment by agnivade** [comment]
Gentle ping @DHaussermann

**Comment by LogFlames** [comment]
Should I try to create a long thread (5000+ replies) and see if I experience the slowdown and issues that I encountered in my installation. Want to make sure to not break things :)

**Comment by DHaussermann** [comment]
Hey @LogFlames I'm taking a look now and not seeing the issue you described. 
image

After a bit of regression testing we should be good to merge this.

**Review by DHaussermann — APPROVED** [review]
Tested and passed 
- Created thread including with 3K+ posts and thread with 6K+ posts
- Open RHS Note there is a delay here of a few seconds with this many replies when open in the channel directly  
- Once thread is opened scrolling and fetching content seems smooth
- Posted several replies. No noticeable delay in typing or posting onto the thread and having the thread reach the server 
- Regression tested no issues or changes in business logic with the thread showing up in Threads view 
- No delays opening the recent activity from threads view 
- Regression tested no changes to when thread becomes unread or has a badge 
- Spot checked with CRT disabled. Opening and using the RHS is comparable

LGTM!

Thanks @agnivade. Sorry for the delay on this.

**Comment by agnivade** [comment]
E2E tests passed. Merging.

**Pull request merged**